py/pin.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016 Paul Sokolovsky
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+
+#define MP_PIN_READ   (1)
+#define MP_PIN_WRITE  (2)
+#define MP_PIN_INPUT  (3)
+#define MP_PIN_OUTPUT (4)
+
+// Pin protocol
+typedef struct _mp_pin_p_t {
+    mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+} mp_pin_p_t;
+
+int mp_pin_read(mp_obj_t pin);
+void mp_pin_write(mp_obj_t pin, int value);
```

Comment:
IMHO yes, as it's focus of uPy. It's also the reason why it's put into py/ instead of extmod/ (this point isn't so criticial of course). Note that mp_pin_* and mp_hal_* are 2 different interfaces. mp_hal_* is port HAL's pin interface, a particular port may either implement it from scratch, or redirect to generic mp_pin_* interface. The latter is thus "internal", and can use less distinguishing naming, suggestions are welcome.